beacon-chain/rpc/beaconv1/validator.go

```diff
@@ -2,27 +2,165 @@ package beaconv1
+			return nil, status.Errorf(codes.Internal, "Could not get balance at index %d: %v", valIndex, err)
+		}
+		valBalances[i] = &ethpb.ValidatorBalance{
+			Index:   valIndex,
+			Balance: bal,
+		}
+	}
+	return &ethpb.ValidatorBalancesResponse{Data: valBalances}, nil
 }
 
 // ListCommittees retrieves the committees for the given state at the given epoch.
+// If the requested slot and index are defined, only those committees are returned.
 func (bs *Server) ListCommittees(ctx context.Context, req *ethpb.StateCommitteesRequest) (*ethpb.StateCommitteesResponse, error) {
-	return nil, errors.New("unimplemented")
+	state, err := bs.StateFetcher.State(ctx, req.StateId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get state: %v", err)
+	}
+
+	epoch := helpers.SlotToEpoch(state.Slot())
+	if req.Epoch != nil {
+		epoch = *req.Epoch
+	}
+	activeCount, err := helpers.ActiveValidatorCount(state, epoch)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get active validator count: %v", err)
+	}
+
+	startSlot, err := helpers.StartSlot(epoch)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get epoch start slot: %v", err)
+	}
+	endSlot, err := helpers.EndSlot(epoch)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get epoch end slot: %v", err)
+	}
+	committeesPerSlot := helpers.SlotCommitteeCount(activeCount)
+	committees := make([]*ethpb.Committee, 0)
+	for slot := startSlot; slot <= endSlot; slot++ {
+		for index := types.CommitteeIndex(0); index < types.CommitteeIndex(committeesPerSlot); index++ {
+			if req.Slot != nil && slot != *req.Slot {
+				continue
+			}
+			if req.Index != nil && index != *req.Index {
+				continue
+			}
+			committee, err := helpers.BeaconCommitteeFromState(state, slot, index)
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "Could not get committee: %v", err)
+			}
+			committeeContainer := &ethpb.Committee{
+				Index:      index,
+				Slot:       slot,
+				Validators: committee,
+			}
+			committees = append(committees, committeeContainer)
+		}
+	}
+	return &ethpb.StateCommitteesResponse{Data: committees}, nil
+}
+
+func valContainerById(state iface.BeaconState, validatorId []byte) (*ethpb.ValidatorContainer, error) {
+	val, valIndex, err := validatorById(state, validatorId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get validator: %v", err)
+	}
+
+	bal, err := state.BalanceAtIndex(valIndex)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get validator balance: %v", err)
+	}
+	valContainer := &ethpb.ValidatorContainer{
+		Index:     valIndex,
+		Balance:   bal,
+		Status:    "",
+		Validator: val,
+	}
+	return valContainer, nil
+}
+
+func validatorById(state iface.BeaconState, validatorId []byte) (*ethpb.Validator, types.ValidatorIndex, error) {
+	var valIndex types.ValidatorIndex
+	if len(validatorId) == params.BeaconConfig().BLSPubkeyLength {
```

Comment:
We should explain what's going on here. I think this function could use some godocs. It is not clear to a reader what a `validatorId` is and we should clarify that it is part of the API request and can refer to either a public key hex string, or a validator index in the beacon state. We can verify which kind it is by checking the length of the input argument

Reply:
Maybe we can call this validatorByRequestId